extras/static_code_analysis.sh

```diff
@@ -0,0 +1,204 @@
+      # Remove the header, the first line.
+      tail -n +2 "${f_base}_cppcheck.txt" > "${f_base}_cppcheck.tmp" && mv "${f_base}_cppcheck.tmp" "${f_base}_cppcheck.txt"
+      if [ -s "${f_base}_cppcheck.txt" ]; then
+        cppcheck_failed=true
+        cat ${f_base}_cppcheck.txt | while read line
+        do
+          print_msg "MSGBLD0155: " "[CPPC] $line"
+        done
+      fi
+      rm ${f_base}_cppcheck.txt     
+      print_msg "MSGBLD0160: " "CPPCHECK for file $f completed."
+
+      # CLANG-FORMAT
+      print_msg "MSGBLD0170: " "Run CLANG-FORMAT for file: $f"
+      # Create a clang-format formatted temporary file and perform a diff with its origin.
+      file_formatted="${f_base}_formatted.txt"
+      file_diff="${f_base}_diff.txt"
+      $CLANG_FORMAT $f > $file_formatted
+      diff $f $file_formatted > $file_diff 2>&1
+      if [ -s "$file_diff" ]; then
+        clang_format_failed=true
+        cat $file_diff | while read line
+        do
+          print_msg "MSGBLD0175: " "[DIFF] $line"
+        done
+      fi
+      rm $file_formatted
+      rm $file_diff
+      print_msg "MSGBLD0180: " "CLANG-FORMAT for file $f completed."
+
+      # Add the file to the list of files with format errors.
+      if $vera_failed || $cppcheck_failed || $clang_format_failed; then
+        format_error_files="$format_error_files $f"
+      fi
+      ;;
+
+    *.py )
+      # PEP8
+      print_msg "MSGBLD0190: " "Run PEP8 check for file: $f"
+      case $f in
+        *user_manual_scripts*)
+          IGNORES=$PEP8_IGNORES_TOPO_MANUAL
+          ;;
+        *examples*)
+          IGNORES=$PEP8_IGNORES_EXAMPLES
+          ;;
+        *)
+          IGNORES=$PEP8_IGNORES
+          ;;
+      esac
+      if ! pep8_result=`$PEP8 --ignore=$PEP8_IGNORES $f` ; then
+        printf '%s\n' "$pep8_result" | while IFS= read -r line
+        do
+          print_msg "MSGBLD0195: " "[PEP8] $line"
+        done
+        # Add the file to the list of files with format errors.
+        format_error_files="$format_error_files $f"
+      fi
+      print_msg "MSGBLD0200: " "PEP8 check for file $f completed."
+      ;;
+
+    *)
+      print_msg "MSGBLD0210: " "File $f is not a C/C++/Python file. Static code analysis and formatting check skipped."
+      continue
+  esac
+done
+
+if $RUNS_ON_TRAVIS; then
+  if [ "x$format_error_files" != "x" ]; then
+    for f in $format_error_files; do
+      print_msg "MSGBLD0220: " "Formatting error in file: $f"
+    done
+  fi
+else
+  if [ "x$format_error_files" != "x" ]; then
+    print_msg "" ""
+    print_msg "" "+ + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + +"
+    print_msg "" "+                 TRAVIS CI MAY REPORT FORMAT ERRORS !                        +"
+    print_msg "" "+ + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + +"
+    for f in $format_error_files; do
+      print_msg "" "Formatting error in file: $f"
+    done
```

Comment:
Commit:  	4588e37
I re-worked the whole section to be able to distinguish between C/C++ and Python files. All errors reported by the tool will also make Travis fail. I deactivated cppcheck and I will add additional command-line options to the tool to turn on and off checks. (Not done yet.) The default for cppcheck will be set to off. This will allow us to easily change the code analysis by toggle defaults, and still gives us the possibility to run additional checks not performed on Travis.